HDFS Snapshot/Restore plugin

costin opened the pull request:
__Associated issue (#15191)__
Migrated from ES-Hadoop. Contains several improvements regarding:

* Security
Takes advantage of the pluggable security in ES 2.2 and uses that in order
to grant the necessary permissions to the Hadoop libs. It relies on a
dedicated DomainCombiner to grant permissions only when needed only to the
libraries installed in the plugin folder

* Testing
Uses a customized Local FS to perform actual integration testing of the
Hadoop stack (and thus to make sure the proper permissions and ACC blocks
are in place) however without requiring extra permissions for testing.
If needed, a MiniDFS cluster is provided (though it requires extra
permissions to bind ports)

* Build system
Picks the build system used in ES (still Gradle)

costin commented:
effectively supersedes #13350

rjernst commented on build.gradle:
```diff
@@ -97,6 +97,7 @@ subprojects {
       // the "value" -quiet is added, separated by a space. This is ok since the javadoc
       // command already adds -quiet, so we are just duplicating it
       // see https://discuss.gradle.org/t/add-custom-javadoc-option-that-does-not-take-an-argument/5959
+      javadoc.options.encoding='UTF8'
```
this should be in BuildPlugin

rjernst replied:
Ooops, nevermind.

costin replied:
That should be a separate PR. On windows, not having that option triggers some encoding exceptions (since Windows replies on its own cp1252 scheme). It can be moved to `BuildPlugin` just as well.

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,184 @@
+    archives distZipHadoop2
+}
+
+test {
```
I think this entire test block can be removed? That system property is no longer used, and the output mode seems to be set for debugging?

costin replied:
Fixed

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,184 @@
+
+integTest {
+    clusterConfig.plugin(pluginProperties.extension.name, zipTree(distZipHadoop1.archivePath))
+    logger.info("Installing plugin " + distZipHadoop1.archivePath)
```
This logging message isn't useful, it would be added at configuration time, and there is already a separate task for each plugin install.

The cluster config for plugin should look like this:
```
integTest {
  cluster {
    plugin esplugin.name, zipTree(distZipHadoop1.archivePath)
  }
}
```

However, I'm concerned that we are making multiple jars here, but not testing them all...

rjernst replied:
My code above assumed it was working (I wrote it before I saw the integ tests were completely disabled below). The plugin command takes a String identifier (just used for task name/output) and a plugin zip file. Given the complexity of how this plugin works, and how the plugin command was set up, I think it will require some playing around to get right. I am fine with leaving the integ test disabled and creating a follow up issue which I can address.

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,184 @@
+    logger.info("Installing plugin " + distZipHadoop1.archivePath)
+}
+
+integTest.enabled = false
```
We should have integration tests...and disabling here negates any configuration above....

costin replied:
I've mentioned this in my initial email.
The tests are all integration tests and are actually executed in the `test` phase. integTest tries to run all the time - I tried configuring it but didn't manage to work it out (hence why it is disabled). I'll try it again with the configuration above and see where that goes.

rjernst replied:
We should have a minimal rest test which just checks the plugin can be installed, like we do for all the other plugins. It is important we have at least this minimal level of checking that things work in a real situation (with a real elasticsearch setup, not mocked up test setup).

rmuir replied:
+1, look for an example in the e.g. cloud plugins.

long term i would consider setting up the mini-hdfs thingy in gradle itself as part of the integ tests process (just like we spin up a bin/elasticsearch) before integ tests and shut it down, so it could be used in rest tests. Then its a real "integ test".

the stuff that runs in "test" is not so realistic. we do our best but it is hard, we don't have a proper plugin structure, we don't have plugins realistically loaded in classloaders, its really not the best environment (but ok for unit tests).

costin replied:
Added a Rest Test and enable back integration testing.

costin commented:
Note the documentation for the plugin can be found [here](https://github.com/elastic/elasticsearch-hadoop/tree/master/repository-hdfs).

In particular note that there are 3 `flavors` of jars offered by this plugin depending on the version of the library used: default (no Hadoop jar embedded, it should be available on the classpath), Hadoop1 and Hadoop2. 
As the plugin needs to work with Hadoop it requires its client libraries - depending on the version or distro used (CDH, HDP, etc...) one can use the already provided builds or simply add their own.
Note that the plugin permissions are tested on Hadoop 1 and 2 and apply only to the jars available under `hadoop-libs` folder. The user is free to customize this as she wants.

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,184 @@
+  testCompile "com.sun.jersey:jersey-core:1.8"
+  
+
+  hadoop1("org.apache.hadoop:hadoop-core:${versions.hadoop1}") {
```
We removed all transitive deps just a couple weeks ago, but it didn't handle other configurations. I think we should change that, so that we are explicit about what we are pulling in (which was the entire point about removing transitive deps). Then the excludes and version forcing below would not be necessary.

costin replied:
The excludes there are also used to use only the minimal number of dependencies for each Hadoop client, simply to keep the plugin zips small. That's what the separate configs are used in the end (they don't leak into the IDE) - in case the dependencies are excluded, it will require the build to include those that are necessary.

rjernst replied:
I understand why the excludes are there. My point was we got rid of this style of using transitive deps + excludes in the rest of elasticsearch. However, I am happy to take care of this myself after you have merged. There is no need for my comment to block this change.

rjernst commented:
Thanks Costin. One thing I still dont' understand: is this one plugin (ie one zip) or is it multiple plugins that must be published separately?

costin commented:
@rjernst The plugin binaries are the same since the API is stable across Hadoop versions. The different zips are just convenience so the user can select whatever _flavor_ she needs. Currently this was handled through qualifiers which means different naming - the plugin name (`repository-hdfs`) is the same but the installed binary is different.

> However, I'm concerned that we are making multiple jars here, but not testing them all...
Valid point. I can create a different test that picks up the hadoop2 jar. In practice actually this means simply running the tests against the `hadoop2` plugin zip.
This works in gradle but I can't see an easy solution for the IDE run (which currently is using Hadoop1 simply because it is smaller and cleaner).

rjernst commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/NoOpTests.java:
```diff
@@ -0,0 +1,10 @@
+public class NoOpTests extends LuceneTestCase {
+
+    public void testDummy() {
+        // mmm, mmm, mmm, mmmhmm
```
Why is this test here??

costin replied:
The `HdfsTests` is marked as third-party and thus doesn't gets executed when executing `gradle build`. Which then fails since there are not tests running - adding this dummy test prevents this from happening ('hey, here's a test').

rmuir replied:
maybe we should fix that? what does it need you to manually setup to run?

rjernst commented on plugins/repository-hdfs/src/test/resources/log4j.properties:
```diff
@@ -0,0 +1,9 @@
+es.logger.level=INFO
```
This file should not be needed, it is in the test-framework.

rjernst commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsTests.java:
```diff
@@ -0,0 +1,217 @@
+ */
+@ThirdParty
+@ClusterScope(scope = Scope.SUITE, numDataNodes = 1, transportClientRatio = 0.0)
+public class HdfsTests extends ESIntegTestCase {
```
Are there any tests other than this single one? It concerns me that it is marked as thirdparty, which means it won't get run in jenkins...

costin replied:
No, not really. Mocking would be complicated and not really useful since the whole reason of the test is to make sure it checks ClassLoading and the permissions. And the mock doesn't do that. I could work on adding some unit tests but they'll not be at the same level.
Maybe the `@ThirdParty` annotation can be removed then to make sure the test runs all the time?
Or once the rest test is added, rely on that instead.
Isn't there a jenkins plan that runs all plugins (all third-party)?

rjernst commented:
This change looks ok from the build side, with the caveats of 2 follow up issues I suggested (which I can take care of once merged). My only concern is the lack of any real test, since the only test I see would only be run as a one off. Could we have some minimal unit test on the HdfsRepository class itself?

rmuir commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/plugin/hadoop/hdfs/HdfsPlugin.java:
```diff
@@ -0,0 +1,195 @@
+                return new AccessControlContext(AccessController.getContext(), new HadoopDomainCombiner());
+            }
+        });
+    }
```
can this be moved to securityutils please? we don't want to expose a lot of this as public

costin replied:
Ooops, need to move it around.

rmuir commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/SecurityUtils.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public static <V> V execute(FileSystem fs, FsCallback<V> callback) throws IOException {
+        try {
```
this method (and a few others here) are security holes without a check. basically, before entering any `doPrivileged` block, please do a check for `org.elasticsearch.SpecialPermission`, see https://www.elastic.co/guide/en/elasticsearch/plugins/2.x/plugin-authors.html#_java_security_permissions

if we just add the check like that consistently and religiously, we just don't have to think about it: we won't have privilege escalation problems.

otherwise without it, if a script for example, gets ahold of this guy and invokes it, we will "forget its a script" because we truncate the stack: then the script can potentially do things it should not be able to do. I hope that makes sense.

rmuir commented on plugins/repository-hdfs/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -0,0 +1,47 @@
+  permission java.lang.RuntimePermission "setContextClassLoader";
+  
+  // Hadoop UGI
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```
what is UGI and why does it do this? Is there any issue in hadoop to fix stuff like this? smells like just messiness :)

costin replied:
UGI = UserGroupInformation
Whenever Hadoop starts (client or not), it tries to figure out the user and in the process, looks at Kerberos even though it is not used.
Code for it [here](https://github.com/apache/hadoop/blob/release-1.2.1/src/core/org/apache/hadoop/security/UserGroupInformation.java#L210).
Nevertheless, the reflection permission is not needed - it's a left over from trying to _hack_ the CodeSource.
Removed it now.

The full permission list (and where they are needed) is listed [here](https://gist.github.com/costin/ef6c0914a9bee70b419f)

costin replied:
After going back and forth (thanks to Eclipse/Gradle cache) it looks like the permission is indeed need. I have updated the permission list (and their cause) [here](https://gist.github.com/costin/ef6c0914a9bee70b419f). In short, it's triggered by the UGI initialization which uses reflection to instantiate some classes:

    at org.apache.hadoop.util.ReflectionUtils.newInstance(ReflectionUtils.java:115)
    at org.apache.hadoop.security.Groups.<init>(Groups.java:48)
    at org.apache.hadoop.security.Groups.getUserToGroupsMappingService(Groups.java:140)
    at org.apache.hadoop.security.UserGroupInformation.initialize(UserGroupInformation.java:206)
    at org.apache.hadoop.security.UserGroupInformation.setConfiguration(UserGroupInformation.java:227)

rmuir replied:
In that case the code seems really bogus, that is just hadoop messiness: I'm not even sure setAccessible or reflection is needed as the class and ctor are public, maybe I am missing something.

rmuir commented on plugins/repository-hdfs/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -0,0 +1,47 @@
+  
+  // JAAS libraries are not loaded with the proper context in Hadoop, hence why the permission is needed here
+  permission java.lang.RuntimePermission "loadLibrary.jaas_nt";
+  
```
is there any way to elaborate on what is happening here? I don't understand the problem, it smells like something in hadoop is really broken here, e.g. jar-helling with the JDK itself? (which we don't check for in jarhell). 

Because jaas should be loaded via extensions mechanism (e.g. its in the jdk itself) which has full privileges.

costin replied:
JAAS is not loaded properly by Hadoop. Will follow-up with more details.

rmuir commented:
i am fine with costin's approach to the security permissions (the evil domaincombiner): he explored a lot of alternatives, and this one is best. Actually pretty creative :) I left a few comments/suggestions.

basically i don't think we should do a lot of cleanup/investigation to the craziness of hadoop before merging the PR in, or even immediately afterwards (though if you already understand the problems, any comments in the source are helpful). 

i would instead prefer after merging that we prioritize improving the integration tests setup (e.g. rest tests) to really realistically simulate the plugin: e.g. get gradle to fire up the HDFS minicluster or some mock of it or what have you.

This way, its easier to clean it up, anyone can try to help, and if tests pass we are good :) Without solid integ tests though, personally I would never want to touch it, i don't have confidence that anything i do wouldn't completely break the plugin.

rmuir commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/SecurityUtils.java:
```diff
@@ -0,0 +1,59 @@
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+
+public class SecurityUtils {
```
we should make this entire class package private and try to contain visibility here as well.

rjernst replied:
Can all the classes here be in the same package too? I don't really see why these handful of classes need a package hierarchy.

costin replied:
Will do. The classes are split across packages to indicate the _public_ (loaded in the Plugin ClassLoader) vs _internal_ (loaded in the Hadoop CL). However further more, the differentiation becomes minimal (it was more about keeping it consistent with the other plugins). However locking down security is definitely better, so I'll pack things together.

dadoonet commented:
Could you also pick up this file in your PR and adjust it if needed? https://github.com/dadoonet/elasticsearch/blob/4ccef59cffdfc956e7f84833b9945fe317b9dd99/docs/plugins/repository-hdfs.asciidoc

costin commented:
@dadoonet Of course. I'll include the docs in the next revision (which should follow up shortly today).

costin commented:
Hi, I have updated the PR (as a separate commit just in case).
I'll add the notes with references to the code based on the previous comments so they don't get hidden by Github (outdated diff)

costin commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/plugin/hadoop/hdfs/HdfsRepositoryRestIT.java:
```diff
@@ -0,0 +1,30 @@
+import com.carrotsearch.randomizedtesting.annotations.Name;
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+public class HdfsRepositoryRestIT extends ESRestTestCase {
```
Rest test. Starts the cluster and installs the plugin.

costin commented on plugins/repository-hdfs/src/test/java/org/elasticsearch/plugin/hadoop/hdfs/UtilsTests.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.elasticsearch.test.ESTestCase;
+
+public class UtilsTests extends ESTestCase {
```
Added an actual unit test instead of the dummy one so even if the integration tests are not running, there are some tests in place and the build can succeed.

costin commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/SecurityUtils.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    static <V> V execute(FileSystem fs, FsCallback<V> callback) throws IOException {
+        SecurityManager sm = System.getSecurityManager();
```
Added check for SpecialPermission before each doPrivileged call. I wonder if going forward, especially with the use of JDK8, we should wrap all ACC calls in a special SecurityUtils class and prevent folks from using it directly through the forbidden api.
Or a shorter solution, use forbidden api anyway when doPrivileged calls are done, just so folks don't forget.

rmuir replied:
Lets please not do that. The thing is it would break: the codebase of the utility class itself would screw everything up. 

To me, the right answer is not to make these acc blocks easier: its to have less of them, by fixing the actual problems so the blocks are not needed :)

costin replied:
Right, forgot about the codebase. Maybe using forbidden api then? That is automate the check for `SpecialPermission` since it's easy to forget about it.

costin commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/plugin/hadoop/hdfs/Utils.java:
```diff
@@ -0,0 +1,84 @@
+
+import org.elasticsearch.SpecialPermission;
+
+public abstract class Utils {
```
The reason for this class existing is so it makes the AccessControlContext of the plugin (based on its codeSource) available to the nested Hadoop classloader (which sits in a different ClassLoader and different codeSource).
The classes need to communicate with each other across ClassLoaders so I went the route of security by obscurity which is, use this class that offers the ACC through a protected method so that the a class in the nested ClassLoader can extend it and thus gain access to its method and ACC context.
In the end this means gaining access to the combiner which checks anyway for the codeSource of the assigned ProtectionDomain so even if other classes would gain access to that, it wouldn't really serve them.

The other alternative was to use split packages - have a class in the same package as HdfsPlugin, but loaded inside the nested Hadoop CL. This way it would not have the privileges of the plugin but it could access to its method (which would be package private instead of protected) and thus gain access to it.
It's more obscure/secure but to the point where folks looking at it might wonder what is going on...

costin commented on plugins/repository-hdfs/src/main/java/org/elasticsearch/plugin/hadoop/hdfs/HdfsPlugin.java:
```diff
@@ -0,0 +1,149 @@
+import org.elasticsearch.repositories.RepositoriesModule;
+import org.elasticsearch.repositories.Repository;
+
+//
```
Added some comments of the whole dance around making Hadoop work properly. Hopefully it is explanatory enough (tried to cover the other alternatives looked at just in case).

costin commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,175 @@
+}
+
+
+dependencyLicenses {
```
This doesn't seem to create any licenses - maybe it is not configured properly?

rjernst replied:
This doesn't creat licenses, it checks them

costin commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -0,0 +1,175 @@
+
+integTest {
+    cluster {
+        plugin(pluginProperties.extension.name, zipTree(distZipHadoop1.archivePath))
```
From what I can tell, this configuration doesn't change the plugin source. The bundlePlugin archive is still used instead, meaning the plugin without the Hadoop jars instead of the one with Hadoop1 in it. Which means the plugin loads but the comment somebody pokes at it, it will fail with ClassNotFoundException (since there are no Hadoop classes available).

costin commented:
Folks, should I merge the branch in master or are there any other things to take of ? Thanks,

rmuir commented:
I'm just repeating my big concern one last time: there are a ton of jars and security permissions needed for this thing, but are the tests sufficient? I only see one test, and its a unit test, and those tests live an in an unrealistic approximation of reality: "fantasy land"

If its pushed as is without test improvements, its reasonable to expect I will break the plugin completely in a matter of days (as we are still aggressively cleaning things up). The only way to avoid this is with tests.

For such situations we need a real integ test, this means gradle spins up HDFSMiniCluster or whatever is needed during `integTest` and there is a rest test that does a snapshot and a restore.

costin commented:
Point taken, I'll follow up with @rjernst to properly configure `integTest` to use the appropriate jar and make sure a snapshot/restore is executed.
For what it's worth, the unit test does apply the permissions - in fact if one is removed from the policy, the test will fail. Also it does perform an actual snapshot and restore (hence why in the initial commit I had logging enabled); what it doesn't do is handle class-loading since the unit test/IDE structure is different from the actual context at runtime.

s1monw commented:
> Point taken, I'll follow up with @rjernst to properly configure integTest to use the appropriate jar and make sure a snapshot/restore is executed.

thanks @costin !!

rmuir commented:
> For what it's worth, the unit test does apply the permissions - in fact if one is removed from the policy, the test will fail.

Yes, this is good, but unit tests do not know which jars are belonging to what thing (just a massive classpath) so it "approximates" and is overly generous: https://github.com/elastic/elasticsearch/blob/master/test-framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java#L175-L178

I agree, the proper integTest is a bigger issue (we need similar for cloud plugins too), it should not block merge to master, but we should open a followup and try to tackle it.

costin commented:
Hi,

I've added another integration test that checks the HDFS repository and delete it. Probably it can be improved (to snapshot an index) but I couldn't find a working example and gave up ...

```
[2015-12-07 23:39:56,805][INFO ][plugins                  ] [Professor X] loaded [repository-hdfs], sites []
[2015-12-07 23:39:56,807][WARN ][plugins                  ] [Professor X] Plugin: repository-hdfs implementing onModule by the type is not of Module type class org.elasticsearch.index.IndexModule
[2015-12-07 23:39:56,831][INFO ][env                      ] [Professor X] using [1] data paths, mounts [[WORK (E:)]], net usable_space [40gb], net total_space [111.7gb], spins? [unknown], types [NTFS]
[2015-12-07 23:39:57,159][INFO ][plugin.hadoop.hdfs       ] Loaded Hadoop [1.2.1] libraries from file:/E:/work/elasticsearch/master/plugins/repository-hdfs/build/cluster/integTest%20node0/elasticsearch-3.0.0-SNAPSHOT/plugins/repository-hdfs/
[2015-12-07 23:39:58,594][INFO ][node                     ] [Professor X] initialized
[2015-12-07 23:39:58,594][INFO ][node                     ] [Professor X] starting ...
[2015-12-07 23:39:58,704][WARN ][common.network           ] [Professor X] publish host: [_local_] resolves to multiple addresses, auto-selecting {127.0.0.1} as single publish address
[2015-12-07 23:39:58,704][INFO ][transport                ] [Professor X] publish_address {127.0.0.1:9500}, bound_addresses {127.0.0.1:9500}, {[::1]:9500}
[2015-12-07 23:39:58,711][INFO ][discovery                ] [Professor X] plugins_repository-hdfs_integTest/1LHITVsLRkGbnG4Qtj5lzg
[2015-12-07 23:40:01,727][INFO ][cluster.service          ] [Professor X] new_master {Professor X}{1LHITVsLRkGbnG4Qtj5lzg}{127.0.0.1}{127.0.0.1:9500}[testattr=>test], reason: zen-disco-join(elected_as_master, [0] joins received)
[2015-12-07 23:40:01,767][INFO ][gateway                  ] [Professor X] recovered [0] indices into cluster_state
[2015-12-07 23:40:01,794][WARN ][common.network           ] [Professor X] publish host: [_local_] resolves to multiple addresses, auto-selecting {127.0.0.1} as single publish address
[2015-12-07 23:40:01,795][INFO ][http                     ] [Professor X] publish_address {127.0.0.1:9400}, bound_addresses {127.0.0.1:9400}, {[::1]:9400}
[2015-12-07 23:40:01,795][INFO ][node                     ] [Professor X] started
[2015-12-07 23:40:08,247][INFO ][repositories             ] [Professor X] put repository [test_repo_hdfs_1]
[2015-12-07 23:40:08,302][INFO ][repositories             ] [Professor X] delete repository [test_repo_hdfs_1]
```

Note that there are several changes made to the build system to make this work as I couldn't figure a better way to do this:

1. `bundlePlugin` configuration cannot be changed since, at least [currently](https://github.com/elastic/elasticsearch/blob/4e80a5e0994f25caa19c3e78573bdc9fa6bcc4c6/buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy#L40),  the Gradle plugin relies on `project.afterEvaluate` which means whatever configuration is added, the plugin comes in last and overrides it.
To overcome this, I have changed the default bundlePlugin to include the hadoop1 libraries and thus, to use that when testing.
Going forward, this should not be the case and a different jar could be specified instead.

2. Instead of starting a MiniHDFSCluster (which is still possible but heavy-weight and can leave threads hanging), I've used the `TestingFs` which is a full-blown local HDFS implementation that runs in process. However I'm not aware of a way to add this jar to the plugin classpath as a separate jar hence why I have moved it into from testing to the main source so it's always included.
Next step would be to move this back to the testing area and configure the build appropriately.

I went for 1 and 2 to have the tests running properly - as both are packaging problems they seem to me less important (though need to be fixed before releasing) then making sure the code is tested constantly.

Let me know what you think,

costin commented on plugins/repository-hdfs/src/test/resources/rest-api-spec/test/hdfs_repository/20_repository.yaml:
```diff
@@ -0,0 +1,25 @@
+              path: "build/data/repo-hdfs"
+
+    # Get repositry
+    - do:
```
Configure and check the HDFS repository. Next step, actually snapshot and restore an index.

dadoonet commented on plugins/repository-hdfs/src/test/resources/rest-api-spec/test/hdfs_repository/20_repository.yaml:
```diff
@@ -0,0 +1,25 @@
+              uri: "es-hdfs://./build/"
+              path: "build/data/repo-hdfs"
+
+    # Get repositry
```
:) s/repositry/repository

rjernst commented on plugins/repository-hdfs/src/test/resources/rest-api-spec/test/hdfs_repository/20_repository.yaml:
```diff
@@ -0,0 +1,25 @@
+            type: hdfs
+            settings:
+              # local HDFS implementation
+              conf.fs.es-hdfs.impl: "org.elasticsearch.repositories.hdfs.TestingFs"
```
This seems like it defeats the purpose of an integration test. ES should be running as real as it would be outside of testing. I understand that we need an easy way to start an associated process with the integration tests (I am working on that), but I don't think we should do this hack, I'm not sure it really tests anything.

costin replied:
As I've mentioned, the reason the `FS` is in there (and yes, it is a real HDFS file-system and a valid one as one can use a local API) is to overcome some of the current build limitations.
It's easy to remove it it and start an over the network HDFS (which will run locally just the same) assuming the build allows it.

Having a test in place, though not perfect, is better in my opinion than having no test as all.

costin commented:
Hi,

Unless somebody objects, I'm going to merge this into master. I'll revert the integration test added last (to address @rjernst comment) so that packaging is not affected. I'll follow up with another ticket to pursue this further.

Once it is merged in master and jenkins picks it up, I'd like to merge it into 2.x/2.2. Any objections? /cc @s1monw @clintongormley (I'll raise a different issue for this one just in case).

costin commented:
I've squashed the PR, rebased it and merged it in master. Will let it sit for a few hours to be picked up by the build system and if everything is fine, follow up with the improvements for Gradle and backport for 2.2.

Thanks everyone for the feedback!
